File: capture_golden.m
%% CAPTURE_GOLDEN Writes reference outputs of the QZ path as plain text
%
%  Copyright 2026 Sustainable Power Systems Lab (SPSL)
%
%  Licensed under the Apache License, Version 2.0 (the 'License');
%  you may not use this file except in compliance with the License.
%  You may obtain a copy of the License at
%
%      http://www.apache.org/licenses/LICENSE-2.0
%
%  Unless required by applicable law or agreed to in writing, software
%  distributed under the License is distributed on an 'AS IS' BASIS,
%  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
%  See the License for the specific language governing permissions and
%  limitations under the License.
%
%  Usage:
%  capture_golden('jac', 'golden')
%  capture_golden('jac', 'golden', 'jac_struc.dat')
%
%  Reads <prefix>_val.dat, <prefix>_eqs.dat and <prefix>_var.dat, runs the
%  reduction and the dense eigensolve, and writes four text files into outdir.
%  This is the reference the Fortran and Python implementations are validated
%  against, so it deliberately avoids analyze_results and loop_analysis: no
%  plotting, no interaction, no base-workspace saving.
%
%  Runs identically under MATLAB and Octave 8. From a shell:
%    octave --no-window-system --quiet --eval \
%      "addpath('..','../scripts'); capture_golden('test','../golden')"
%
%  Outputs, all 1-based and full double precision:
%    <name>_Asys.txt    i j value, one nonzero of the state matrix per line
%    <name>_eigs.txt    real imag, one eigenvalue per line
%    <name>_pf.txt      dense participation factors, rows = differential
%                       states, columns = modes, normalised so each column
%                       peaks at 1
%    <name>_states.txt  index family device variable, one differential state
%                       per line, giving the row labels of _pf.txt
%
%  Modes are sorted by descending real part, then descending imaginary part,
%  so the files are reproducible. Note that eigenvectors are NOT captured:
%  their scale and phase are arbitrary, and for degenerate eigenvalues they
%  are not even unique, so they cannot serve as a reference. Participation
%  factors are only meaningful for simple eigenvalues for the same reason.

function capture_golden(prefix, outdir, jac_struc)

global dif_states raw_vars Jdyn Jdyn_sp

if nargin < 3 || isempty(jac_struc)
    jac_struc = '';
end
if nargin < 2 || isempty(outdir)
    error('capture_golden requires a prefix and an output directory.');
end

[~, name] = fileparts(prefix);

% init rejects a non-finite Jacobian, so a NaN export fails here with a clear
% message rather than producing an all-NaN spectrum.
init([prefix '_val.dat'], [prefix '_eqs.dat'], [prefix '_var.dat'], jac_struc);
calc_Jdyn();

[V, D, W] = eig(Jdyn);
lam = diag(D);

%% Deterministic order: descending real part, then descending imaginary part
[~, ord] = sortrows([real(lam), imag(lam)], [-1 -2]);
lam = lam(ord); V = V(:,ord); W = W(:,ord);

%% Participation factors, as analyze_results.m computes them
P = abs(W .* V);
for i = 1:size(P,2)
    P(:,i) = P(:,i) / max(P(:,i));
end

%% Residual check: independent of any reference, so it must hold here
res = norm(Jdyn*V - V*diag(lam), 'fro') / norm(Jdyn, 'fro');
fprintf('%s: Nx=%d, relative eigenpair residual %.3e\n', name, numel(lam), res);
if res > 1e-10
    error('Eigenpair residual %.3e is too large to use as a reference.', res);
end

if ~exist(outdir, 'dir')
    mkdir(outdir);
end

[ri, ci, vi] = find(Jdyn_sp);
write_rows([outdir filesep name '_Asys.txt'], '%d %d %.17e\n', [ri(:), ci(:), vi(:)]);
write_rows([outdir filesep name '_eigs.txt'], '%.17e %.17e\n', [real(lam), imag(lam)]);

fid = fopen([outdir filesep name '_pf.txt'], 'w');
fprintf(fid, [repmat('%.17e ', 1, size(P,2)) '\n'], P.');
fclose(fid);

fid = fopen([outdir filesep name '_states.txt'], 'w');
for k = 1:numel(dif_states)
    fprintf(fid, '%d %s %s %s\n', dif_states(k), ...
            char(raw_vars{3}(dif_states(k))), ...
            char(raw_vars{4}(dif_states(k))), ...
            char(raw_vars{5}(dif_states(k))));
end
fclose(fid);

fprintf('%s: wrote 4 golden files to %s\n', name, outdir);
end

function write_rows(fname, fmt, data)
fid = fopen(fname, 'w');
fprintf(fid, fmt, data.');
fclose(fid);
end

File: capture_golden.py
#!/usr/bin/env python3
"""Capture reference outputs of the small-signal reduction as plain text.

Reads an exported Jacobian (``<prefix>_val.dat``, ``_eqs.dat``, ``_var.dat``),
reduces it to the state matrix, solves the dense eigenproblem, and writes the
reference files that ``tests/test_golden.py`` and the RAMSES release gates are
checked against.

This is deliberately an **independent** implementation. It shares no code with
the Fortran engine it validates, so agreement between the two is evidence
rather than a tautology. It is a direct port of the retired MATLAB reference
implementation and reproduces its output to better than 1e-14.

Usage::

    python capture_golden.py fixtures/kundur_pss golden
    python capture_golden.py fixtures/test golden

Outputs, all 1-based and full double precision:

``<name>_Asys.txt``
    ``i j value`` per nonzero of the state matrix. Captured separately from the
    eigenvalues on purpose: it separates a reduction error from an eigensolve
    error, which one comparison alone cannot distinguish.
``<name>_eigs.txt``
    ``real imag`` per eigenvalue.
``<name>_pf.txt``
    Dense participation factors, rows = differential states, columns = modes,
    each column normalised to peak at 1.
``<name>_states.txt``
    ``index family device variable`` per differential state, giving the row
    labels of ``_pf.txt``.

Modes are sorted by descending real part, then descending imaginary part, so the
files are reproducible.

Eigenvectors are **not** captured. Their scale and phase are arbitrary, and for
a repeated eigenvalue they are not unique at all, so they cannot serve as a
reference. Participation factors are only comparable for simple eigenvalues for
the same reason: identical device models produce identical poles, and these
systems are heavily degenerate.
"""

import os
import sys

import numpy as np
from scipy.linalg import eig
from scipy.sparse import coo_matrix
from scipy.sparse.linalg import splu


def build_state_matrix(prefix):
    """Return the state matrix, the differential state indices, and their labels.

    The linearised model is ``dx/dt = fx dx + fy dy`` and ``0 = gx dx + gy dy``.
    Eliminating the algebraic variables gives the Schur complement
    ``A_sys = fx - fy gy^-1 gx``, which exists only if ``gy`` is nonsingular.
    """
    val = np.loadtxt(prefix + "_val.dat")
    if val.ndim != 2 or val.shape[1] < 3:
        raise SystemExit(
            "%s_val.dat is not a plain 3-column coordinate file. Non-numeric "
            "entries such as NaN in the value column are the usual cause." % prefix)
    if not np.isfinite(val[:, 2]).all():
        raise SystemExit(
            "%s_val.dat holds %d non-finite entries (NaN or Inf) of %d; the "
            "spectrum would be all NaN."
            % (prefix, int((~np.isfinite(val[:, 2])).sum()), val.shape[0]))

    n = int(max(val[:, 0].max(), val[:, 1].max()))
    # coo -> csr sums duplicate entries, which the assembly does produce.
    S = coo_matrix((val[:, 2], (val[:, 0].astype(int) - 1, val[:, 1].astype(int) - 1)),
                   shape=(n, n)).tocsr()

    gamma = np.loadtxt(prefix + "_eqs.dat", usecols=5, dtype=int)
    names = np.loadtxt(prefix + "_var.dat", usecols=(1, 2, 3, 4), dtype=str)

    # Permute rows so equation i differentiates state i, making E diagonal. The
    # algebraic rows land in whatever order remains, which does not matter: gy
    # and gx receive the same permutation and (P gy)^-1 (P gx) = gy^-1 gx.
    perm = np.empty(n, dtype=int)
    claimed = np.zeros(n, dtype=bool)
    for j, g in enumerate(gamma):
        if g != 0:
            perm[g - 1] = j
            claimed[g - 1] = True
    free = [j for j, g in enumerate(gamma) if g == 0]
    for slot, j in zip(np.flatnonzero(~claimed), free):
        perm[slot] = j
    S = S[perm, :]

    dif = np.flatnonzero(names[:, 0] == "d")
    alg = np.flatnonzero(names[:, 0] != "d")

    fx, fy = S[dif][:, dif], S[dif][:, alg]
    gx, gy = S[alg][:, dif], S[alg][:, alg]

    # Factor gy once and back-substitute, rather than re-solving per column.
    gygx = splu(gy.tocsc()).solve(gx.toarray())
    a_sys = np.asarray((fx - fy @ gygx).todense()) if hasattr(fx - fy @ gygx, "todense") \
        else np.asarray(fx - fy @ gygx)
    return a_sys, dif, names


def capture(prefix, outdir):
    name = os.path.basename(prefix)
    a_sys, dif, names = build_state_matrix(prefix)
    nx = a_sys.shape[0]

    lam, vl, vr = eig(a_sys, left=True, right=True)

    order = np.lexsort((-lam.imag, -lam.real))
    lam, vl, vr = lam[order], vl[:, order], vr[:, order]

    # Participation, normalised per mode so the arbitrary eigenvector scaling
    # cancels. abs() is taken before normalising, matching the reference.
    pf = np.abs(np.conj(vl) * vr)
    pf /= pf.max(axis=0)

    # Independent of any reference, so it must hold here.
    res = np.linalg.norm(a_sys @ vr - vr * lam, "fro") / np.linalg.norm(a_sys, "fro")
    print("%s: Nx=%d, relative eigenpair residual %.3e" % (name, nx, res))
    if res > 1e-10:
        raise SystemExit(
            "Eigenpair residual %.3e is too large to use as a reference." % res)

    os.makedirs(outdir, exist_ok=True)
    j, i = np.meshgrid(np.arange(nx), np.arange(nx))
    nz = a_sys != 0.0
    np.savetxt(os.path.join(outdir, name + "_Asys.txt"),
               np.column_stack([i[nz] + 1, j[nz] + 1, a_sys[nz]]),
               fmt=["%d", "%d", "%.17e"])
    np.savetxt(os.path.join(outdir, name + "_eigs.txt"),
               np.column_stack([lam.real, lam.imag]), fmt="%.17e")
    np.savetxt(os.path.join(outdir, name + "_pf.txt"), pf, fmt="%.17e")
    with open(os.path.join(outdir, name + "_states.txt"), "w") as fh:
        for k in dif:
            fh.write("%d %s %s %s\n" % (k + 1, names[k, 1], names[k, 2], names[k, 3]))

    print("%s: wrote 4 golden files to %s" % (name, outdir))


if __name__ == "__main__":
    if len(sys.argv) != 3:
        raise SystemExit(__doc__)
    capture(sys.argv[1], sys.argv[2])
